### ISSUE TITLE
Authenticating and searching an LDAP domain server via TLS or ldapi

### ISSUE THREAD

**Issue by rdmark** [Netatalk/netatalk#762]
### Discussed in https://github.com/Netatalk/netatalk/discussions/761

<div type='discussions-op-text'>

<sup>Originally posted by **UUVE** March 28, 2024</sup>
As the title suggests, a very useful addition would be the ability to authenticate to an LDAP server and look up a configured domain via the TLS or ldapi (unix domain socket) protocols. For security reasons I would prefer to start my openLDAP server only accepting connections via TLS. The ldapi protocol would then be useful for situations where the ldap server and the netatalk server run on the same machine.</div>

**Comment by rdmark** [comment]
Cross-posting from the discussion:

How is netatalk currently authenticating to an LDAP server... http? I looked at the netatalk code and could only conclude that we don't do any authentication handling on our end. Rather, I anticipate this is done through the LDAP library API, i.e. when calling `ldap_init`.

In fact, as I researched this I found https://github.com/Netatalk/netatalk/pull/11 where the LDAP API init call was modernized to support more protocols, but then reverted again because it broke compatibility with Solaris. It's been 9 years since that event, and Solaris has seen a handful of updates. I say we revisit this decision.

**Comment by rdmark** [comment]
@UUVE Would you be able to test the code in this PR? https://github.com/Netatalk/netatalk/pull/768

I've never set up my own LDAP server or used LDAP with netatalk, so there's a bit of a hurdle for me to test it myself. Your support would be appreciated!

**Comment by UUVE** [comment]
@rdmark Ok, I'll try to recompile my debian package soon (latest version 3.1.8 of Netatalk)

**Comment by UUVE** [comment]
... with the patch applied of course

**Comment by UUVE** [comment]
Is there an easy method on Github to download a full commit patch against Netatalk version 3.1.18, or can you post such a patch to speed up my testing?

**Comment by UUVE** [comment]
My request is no more necessary. Already patched and compiled the source code. Now testing

**Comment by UUVE** [comment]
Netatalk start normally and from system log seems to query the openLDAP server. However afpldaptest reports an error of user not found:
`root@server1.home:~/software/netatalk/netatalk-3.1.18-2_build# afpldaptest -u test
Start parsing afp.conf
Mar 29 18:38:22.437250 [2755963] {ldap_config.c:103} (info:AFPDaemon): LDAP: not used
Finished parsing afp.conf
afp.conf is not ok, not using LDAP. Only local UUID testing available.
Searching user: test
Mar 29 18:38:22.439836 [2755963] {uuid.c:184} (error:AFPDaemon): getuuidfromname("test",t:3693937911): unknown user
User test not found.`
Maybe we should also patch afpldaptest?

**Comment by UUVE** [comment]
As a reminder: the uri ldapi./// requires, with default access rules, that the user connecting to the openLDAP server is the root user. I don't know how this may impact the netatalk server and have not currently investigated whether it is possible to add access rules to allow other users to connect to the openLDAP local socket.

**Comment by UUVE** [comment]
Sorry! My mistake in writing the ldap preferences in afp.conf! I have waited ldap_uri = ... instead of ldap uri = ...
Now is working!
`root@server1.home:~/software/netatalk/netatalk-3.1.18-2_build# afpldaptest -u testuser
Start parsing afp.conf
Finished parsing afp.conf
afp.conf is ok. Using simple bind.
Searching user: testuser
User: testuser ==> UUID: CA621AEA-5723-453A-A330-D6087184016B
root@server1.home:~/software/netatalk/netatalk-3.1.18-2_build#`

**Comment by UUVE** [comment]
I can confirm that it's working passing all the three types of uri in afp.conf:
`ldap uri = ldapi:///
ldap uri = ldap://localhost
ldap uri = ldaps://localhost`

**Comment by UUVE** [comment]
> As a reminder: the uri ldapi./// requires, with default access rules, that the user connecting to the openLDAP server is the root user. I don't know how this may impact the netatalk server and have not currently investigated whether it is possible to add access rules to allow other users to connect to the openLDAP local socket.
This is not true, sorry, the default openldap acls is that the root user has complete access to the directory tree and its configuration via SASL EXTERNAL authentication.

**Comment by rdmark** [comment]
@UUVE Thanks for testing. So the conclusion is that this patch passed all your tests, correct?

There's one more test case that I wanted to ask you about. The [LDAP API documentation](https://linux.die.net/man/3/ldap_initialize) suggests there is a fourth supported protocol for the URI called `cldap` (connectionless LDAP). Are you familiar with this, and would you be able to test if it works with netatalk?

**Comment by UUVE** [comment]
For the tests I used afplapdtest for all connection types, always obtaining correct results in finding the UUID attribute of the users present in the directory. For me the test can be considered concluded. I believe the update break the connection to an LDAP server of Netatalk installations that use it. In fact, you must replace in afp.conf the "ldap server" preference with the "ldap uri" preference which has a different syntax (ldap uri = ldap://somehost or ldap uri = ldaps://somehost or ldap uri = ldapi:///, the latter for connections to an ldap server running on the same machine as Netatalk). This fact should be clearly written in the release notes as well as in the documentation.
As for the cldap protocol, I didn't know it, however searching on the internet it seems to be considered historical and deprecated by RFC 3352 of 2003 (https://datatracker.ietf.org/doc/rfc3352/)

**Comment by rdmark** [comment]
@UUVE Got it, then I will not mention the `cldap` protocol in the netatalk docs.

Well noted that this is a breaking change for existing deployments of netatalk. In addition to the change in the afp.conf option, we will also require a somewhat more recent version of the LDAP library (e.g. openldap). While running our test suite, I found out that OmniOS ships with a too old version (seems like 2.0) by default. My research suggests that at least v2.3 is required for `ldap_initialize`. But luckily, the ooce package repo has an optional openldap 2.6.7 package that is now confirmed working (or at least linking) with netatalk with this patch applied.

So in conclusion, we introduce two breaking changes with this patch. It is good timing, however, since we are preparing to release v3.2.0 of netatalk, which already has a number of breaking changes. So I will make sure we list this with all the other big changes in the release notes.

**Comment by UUVE** [comment]
Perfect! In my eyes this looks like a big improvement.

**Comment by rdmark** [comment]
Merged to both main and 2.4 branches.
